nativepython/tests/string_compilation_test.py

```diff
@@ -457,3 +457,67 @@ def test_find_perf(f, s, subs):
                 .format(c_elapsed, elapsed)
         )
         """
+
+    def check_string_join(self, f):
+        """This function contains tests for implementation of the string join
+        for both: ListOf(str) and TupleOf(str).
+        The tests for the types are as separate functions, so it will be easier
+        to find the failing line later.
+        """
+        res = f(",", ListOf(str)(["1", "2", "3"]))
+        self.assertEqual("1,2,3", res)
+
+        # test longer separator
+        res = f("---", ListOf(str)(["1", "2", "3"]))
+        self.assertEqual("1---2---3", res)
+
+        # test longer items
+        res = f("---", ListOf(str)(["aaa", "bb", "c"]))
+        self.assertEqual("aaa---bb---c", res)
+
+        # test empty separator
+        res = f("", ListOf(str)(["1", "2", "3"]))
+        self.assertEqual("123", res)
+
+        # test empty list
+        res = f(",", ListOf(str)([]))
+        self.assertEqual("", res)
+
+        # test empty list with empty separator
+        res = f("", ListOf(str)([]))
+        self.assertEqual("", res)
+
+    def test_string_join_for_tuple_of_str(self):
+        # test passing tuple of strings
+        @Compiled
+        def f_tup_str(sep: str, items: TupleOf(str)) -> str:
+            return sep.join(items)
+
+        self.check_string_join(f_tup_str)
+
+    def test_string_join_for_list_of_str(self):
+
+        @Compiled
+        def f(sep: str, items: ListOf(str)) -> str:
+            return sep.join(items)
+
+        self.check_string_join(f)
+
+    def test_string_join_for_bad_types(self):
+        """str.join supports only joining ListOf(str) or TupleOf(str)."""
+
+        # test passing tuple of ints
+        @Compiled
+        def f_tup_int(sep: str, items: TupleOf(int)) -> str:
+            return sep.join(items)
+
+        with self.assertRaisesRegex(TypeError, ""):
+            f_tup_int(",", ListOf(int)([1, 2, 3]))
+
+        # test passing list of other types than strings
+        @Compiled
+        def f_int(sep: str, items: ListOf(int)) -> str:
+            return sep.join(items)
+
+        with self.assertRaisesRegex(TypeError, ""):
+            f_int(",", ListOf(int)([1, 2, 3]))
```

Comment:
we should check that this also works on a Dict(str, str) or a ConstDict(str, str), since both of those are currently iterable, and we need to verify that theywork.

Reply:
done